[STN-60] Multi-theme Sass build

nrambeck opened the pull request:
# Summary

Configures a Sass build in the Humsci Basic theme that generates different stylesheets for each of the sub-themes.

- Uses theme specific Grunt task to run Sass and PostCSS tasks on Sass files in the theme
- NPM scripts are setup to run the Grunt tasks (`npm run watch` and `npm run build:sass`)
- The build generates a stylesheet for each of the 3 themes in the `dist/css` folder
- The 3 Drupal themes are configured to load their respective stylesheets found in Humsci Basic
- linting is also set up as npm script (`npm run lint:sass`) and will be run during the watch task at each change
- sass-true tests have been set up with mocha (`npm run test:sass`)

# TODO

-  ✅ We should setup a Grunt task for linting the Sass stylesheets: https://stylelint.io/
- ✅ Setup mixin that would allow adding styles for multiple themes
- ✅ play around with a card to do Adam's version of set up with sass themes in subthemes https://github.com/SU-HSDO/suhumsci/pull/438
- ✅ set up true Sass tests

# Need Review By (Date)
_['10/30', 'asap', etc.]_

# Urgency
_['low', 'medium', 'high', etc.]_

# Steps to Test
Make some changes to the sass files in humsci_basic and see if these scripts work with them:

- `npm run build:css` should build your new sass to css and apply postCss where necessary
- `npm run lint:sass` should check your new sass for errors (add an error if you'd like)
- `npm run watch` will start watching the humsci_basic src/scss folder. Make a change and see it run.
- `npm run test:sass` make sure sass tests pass

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

Commit:
STN-60:feat: Setup Grunt build for Humsci Basic styles

- also configured subthemes to use stylesheets from humsci_basic

nrambeck commented on docroot/themes/humsci/humsci_basic/src/scss/_base.scss:
```diff
@@ -0,0 +1,18 @@
+  // TODO: Create mixin to support this
+  //@include theme(colorful, airy) {
+  //  background-color: gray;
+  //}
```
Example usage of a mixin that takes multiple themes as parameters. We still need to create the mixin that would do this.

silverli commented on docroot/themes/humsci/humsci_basic/.gitignore:
```diff
@@ -0,0 +1 @@
+dist/
```
would it be better to keep this in the main .gitignore?

nrambeck replied:
Good call. It may be better to keep it all in the main `.gitignore` to avoid any confusion.

silverli replied:
hmm now do we want to keep the css artifact like the rest of the repo though?

silverli replied:
let's delete this ignore for now

Commit:
STN-60:feat: adds stylelint, multi-theme mixin, and sass-true specs

- also cleans up scss naming
- adds grunt run for running stylelint task with grunt
- adds .stylelintrc based on Sparkbox standards and config
- you can now call themes(('')) to add styles for multiple themes
- docs: adds comments on mixin use
- sass-true config, and tests/ folder set up

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/airy.scss:
```diff
@@ -0,0 +1,3 @@
+$current-theme: 'airy';
```
I'd like to play around with CSS variables a bit too to see if they give us anything else

silverli replied:
defer this to next css work

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -0,0 +1,84 @@
+          {
+            expand: true,
+            cwd: 'src',
+            src: ['scss/**/[a-z]*.scss'],
```
ohh so placing this under the src folder negates it from the root gruntfile command right?

nrambeck replied:
If you are talking about the `Gruntfile` in the root of the repository, then no, this would not negate that at all. We would need to change the root Gruntfile somehow to skip over this theme (and the sub-themes too probably).

silverli replied:
Nathan's PR for decanter does this.

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -0,0 +1,84 @@
+    watch: {
+      css: {
+        files: ['**/*.{scss,sass}'],
+        tasks: ['sass', 'run:stylelint'], // run stylelint at same time after each change
```
we need to exclude this scss folder from the root grunt task

sherakama commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -0,0 +1,84 @@
+  function getIncludeFiles() {
+    const patterns = [
+      '**/*.scss',
+      'node_modules/decanter/scss',
```
You'll likely want to include a path to the node_modules directory as well. Future releases of Decanter and Stanford Basic are going to standardize on full paths relative to the node_modules directory for anything brought in through NPM.

silverli replied:
Great! Thank you @sherakama I think Nathan implemented these changes in his new PR here: https://github.com/SU-HSDO/suhumsci/pull/439

sherakama commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -0,0 +1,84 @@
+      options: {
+        map: true, // inline sourcemaps
+        processors: [
+          require('autoprefixer')({
```
FYI our browserlist config: https://github.com/SU-SWS/stanford_basic/blob/8.x-4.x/.browserslistrc

silverli replied:
got it 👍  this is defined in our package.json

sherakama commented on docroot/themes/humsci/humsci_basic/src/scss/_base.scss:
```diff
@@ -0,0 +1,19 @@
+    color: yellow;
+  }
+
+  @include themes(('traditional', 'airy')) {
```
Neat.

silverli replied:
🎉

silverli commented on Gruntfile.js:
```diff
@@ -65,7 +65,7 @@ module.exports = function (grunt) {
             expand: true,
             cwd: 'docroot',
-            src: ['modules/humsci/**/[a-z]*.scss', 'themes/humsci/**/[a-z]*.scss'],
+            src: ['modules/humsci/**/[a-z]*.scss', 'themes/humsci/**/[a-z]*.scss', '!themes/humsci/humsci_basic/**/[a-z]*.scss'],
```
this doesnt seem to ignore changes to the sass-specs folder in humsci_basic

silverli replied:
Nathan's work addresses this https://github.com/SU-HSDO/suhumsci/pull/439

asimpson commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -0,0 +1,84 @@
+    watch: {
+      css: {
+        files: ['**/*.{scss,sass}'],
+        tasks: ['sass', 'run:stylelint'], // run stylelint at same time after each change
```
running stylelint after every change is going to be slow, better to offload that task to an editor plugin and run stylelint in CI.

silverli replied:
ok. let's remove this from our grunt watch task and make a card to add sass-tests and linting to the CI for this repo

nrambeck commented on docroot/themes/humsci/humsci_basic/package.json:
```diff
@@ -0,0 +1,32 @@
+    "test:sass": "mocha 'true-sass-spec.js'",
+    "test": "npm run lint:sass && npm run test:sass"
+  },
+  "author": "",
```
Should we just delete this?

silverli replied:
sounds good to me!

nrambeck commented on docroot/themes/humsci/humsci_basic/sass-specs/_tools.mixins.themes.scss:
```diff
@@ -0,0 +1,133 @@
+      }
+    }
+  }
+}
```
Great test coverage with these tests! I also love the naming convention to match the file name of those mixins being tested. Could we add an empty line at the end of the file? Linters usually complain about that.

silverli replied:
I might need to make sure this is getting linted as well

silverli replied:
hmm actually scss rules for our sass-true tests don't align, so I won't include that in the lint task

nrambeck commented on docroot/themes/humsci/humsci_basic/sass-specs/sass-tests.scss:
```diff
@@ -0,0 +1,11 @@
+// Import all test partials here
+@include describe('Sass Unit Tests') {
+  @import 'tools.mixins.themes';
+}
```
Great setup! What do you think about moving the entire `sass-specs` folder into a tests directory? I'm thinking we could have JS tests in the future, and it would be nice to have them all organized together under one folder.

silverli replied:
good idea!

nrambeck commented on docroot/themes/humsci/humsci_basic/src/scss/_tools.mixins.themes.scss:
```diff
@@ -0,0 +1,50 @@
+// E.g.
+// @themes(('airy', 'colorful'))
+// This has a default to fallback to applying to all themes if nothing is passed in
+@mixin themes($theme-list: $global-theme-list) {
```
I think there is a case to be made to drop the default value for this parameter, so that if no arguments are included with the mixin, it will results in an error. I'm thinking that if I'm using this mixin and don't include any arguments, it's probably not intentional since I would have just added those styles without a mixin at all. If we remove the default value, an error will be thrown in that case letting us know something is amiss.

silverli replied:
ok! that sounds fair :) I am realizing too, that since sass is a dev side library, we need to know if there are errors before css compilation anyways so might as well make it error and not fail silently (I was thinking more in line with JS resiliency before)

silverli replied:
I've changed this and the relevants tests

nrambeck reviewed (commented):
Great work on this! I left just a few inline comments.
